lab_3_generate_by_ngrams/main.py

```diff
@@ -388,6 +624,42 @@ def run(self, prompt: str, seq_len: int) -> Optional[str]:
         In case of corrupt input arguments or methods used return None,
         None is returned
         """
+        if not isinstance(seq_len, int) or seq_len <= 0 or not isinstance(prompt, str) or not prompt:
+            return None
+
+        encoded_prompt = self._text_processor.encode(prompt)
+        if encoded_prompt is None:
+            return None
+        candidates = {encoded_prompt: 0.0}
+
+        updated_candidates = candidates.copy()
+
+        for _ in range(seq_len):
+            candidates = updated_candidates
+            not_sorted_candidates = {}
+
+            for sequence in candidates:
+                next_tokens = self._get_next_token(sequence)
+                if not next_tokens:
+                    return None
+
+                possible_sequences = self.beam_searcher.continue_sequence(sequence, next_tokens, updated_candidates)
+                if not possible_sequences:
+                    return self._text_processor.decode(sorted(tuple(updated_candidates), key=lambda x: x[1])[0])
```

Comment:
I don't understand why breaking a cycle would hurt

Reply:
If I break this cycle, not_sorted_candidates dict is empty, so method prune_sequence_candidates make this whole method return None when none_continue_sequences test expects decoded prompt to be returned. I tried to also breaking a cycle if prune_sequence_candidates return None but in that case this whole method would return decoded prompt when none_prune_sequence_candidates test expects None to be returned.